python/ray/data/datasource/datasink.py

```diff
@@ -26,20 +29,16 @@ def write(
         self,
         blocks: Iterable[Block],
         ctx: TaskContext,
-    ) -> Any:
+    ) -> None:
         """Write blocks. This is used by a single write task.
 
         Args:
             blocks: Generator of data blocks.
             ctx: ``TaskContext`` for the write task.
-
-        Returns:
-            A user-defined output. Can be anything, and the returned value is passed to
-            :meth:`~ray.data.Datasink.on_write_complete`.
         """
         raise NotImplementedError
 
-    def on_write_complete(self, write_results: List[Any]) -> None:
+    def on_write_complete(self, raw_write_results: List[Block]) -> None:
```

Comment:
Nit: By directly passing the raw write results, this is coupling `Datasink` with the implementation of `generate_collect_write_stats_fn` (e.g., you need to read the code of `generate_collect_write_stats_fn` to understand what each block represents.

I think we could decouple this by introducing an abstraction like a `WriteResult` dataclass and changing the type to `List[WriteResult`].

Reply:
I looked deeper into the map codepath, and I think it would be very cumbersome for the map transform to return anything other than another block (there are numerous type checks, getting BlockMetadata, etc). so for now, i think the best option is to just return a pandas dataframe from the maptransform.

Reply:
Was able to hack around it by wrapping the `WriteResult` object in a pandas dataframe.